src/aiida/orm/fields.py

```diff
@@ -0,0 +1,297 @@
+###########################################################################
+"""Module which provides decorators for AiiDA ORM entity -> DB field mappings."""
+from abc import ABCMeta
+from copy import deepcopy
+from pprint import pformat
+from typing import Any, Dict, Iterable, Optional, Sequence, Tuple, Type, Union
+
+__all__ = ('QbAttrField', 'QbField', 'QbFields', 'QbFieldFilters')
+
+
+class QbField:
+    """A field of an ORM entity, accessible via the ``QueryBuilder``"""
+
+    __slots__ = ('_key', '_qb_field', '_doc', '_dtype', '_subscriptable')
+
+    def __init__(
+        self,
+        key: str,
+        qb_field: Optional[str] = None,
+        *,
+        dtype: Optional[Type[Any]] = None,
+        doc: str = '',
+        subscriptable: bool = False,
+    ) -> None:
+        """Initialise a ORM entity field, accessible via the ``QueryBuilder``
+
+        :param key: The key of the field on the ORM entity
+        :param qb_field: The name of the field in the QueryBuilder, if not equal to ``key``
+        :param dtype: The data type of the field. If None, the field is of variable type.
+        :param doc: A docstring for the field
+        :param subscriptable: If True, a new field can be created by ``field["subkey"]``
+        """
+        self._key = key
+        self._qb_field = qb_field if qb_field is not None else key
+        self._doc = doc
+        self._dtype = dtype
+        self._subscriptable = subscriptable
+
+    @property
+    def key(self) -> str:
+        return self._key
+
+    @property
+    def qb_field(self) -> str:
+        return self._qb_field
+
+    @property
+    def doc(self) -> str:
+        return self._doc
+
+    @property
+    def dtype(self) -> Optional[Type[Any]]:
+        return self._dtype
+
+    @property
+    def subscriptable(self) -> bool:
+        return self._subscriptable
+
+    def _repr_type(self, value: Any) -> str:
+        """Return a string representation of the type of the value"""
+        if value == type(None):
+            return 'None'
+        if isinstance(value, type):
+            # basic types like int, str, etc.
+            return value.__qualname__
+        if hasattr(value, '__origin__') and value.__origin__ == Union:
+            return 'typing.Union[' + ','.join(self._repr_type(t) for t in value.__args__) + ']'
+        return str(value)
+
+    def __repr__(self) -> str:
+        dtype = self._repr_type(self.dtype) if self.dtype else ''
+        return (
+            f'{self.__class__.__name__}({self.key!r}'
+            + (f', {self._qb_field!r}' if self._qb_field != self.key else '')
+            + (f', dtype={dtype}' if self.dtype else '')
+            + (f', subscriptable={self.subscriptable!r}' if self.subscriptable else '')
+            + ')'
+        )
+
+    def __str__(self) -> str:
+        type_str = (
+            '?' if self.dtype is None else (self.dtype.__name__ if isinstance(self.dtype, type) else str(self.dtype))
+        )
```

Comment:
This `if` statement is upsetting mypy, specifically raising a `redundant-expr` error, as `if isinstance(whatever, type)` is always `True`. Or is it not? Can't quite tell what @chrisjsewell was aiming for with this.

Reply:
I think so:
```
In [1]: isinstance(object, type)
Out[1]: True
```
and since I think pretty much everything is an `object` the conditional is a bit redundant

Reply:
Right. Odd to think Chris would make this mistake though. Suspicious. Perhaps there was a reason? Can't see one though.

Reply:
There might indeed be an edge case. I am not sure which approach is "safer" to always work, calling `.__name__` or `str()`. I wonder in which cases either one or both won't work.

Reply:
Will take another look at it later. It's one of many `mypy` complaints. Going for the low hanging fruit first.